### PR TITLE
Support logged batches in GraphQL

### LINKED ISSUE #133: Add batching to mutations
Atomicity can be important for some applications. Since batches has additional limitations to be used successfully, we should provide an alternative path for mutations that can be batched.

Batched statements have the following restrictions:

No light weight transactions

No mix of batched and unbatched statements

Avoid the word 'batch' as it could be unclear of the desired use case

No range queries

```
# Atomic directives override the serial execution order of mutations and allow 
# mutations to be executed in an all-or-nothing manner. IfExists, IfNotExists, 
# and ranged partition updates and deletes are forbidden.
directive @atomic

mutation @atomic {
  insertCart(value:{cartId: "..."}) {
    applied
  }
  insertUser(value:{userId: "..."}) {
    applied
  }
}
```

Alternative we can redefine all functions with an atomic_ prefix that only allow sane use of atomic statements. This would be safer but adds a lot of extra definitions to the schema.

### PR DESCRIPTION
**Pull request by jorgebay**
Resolves #133 .

It should be rebased once #271 and #279 are merged.

### CODE REVIEW

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java  5
- testing/src/test/java/io/stargate/it/http/queries/GraphQLAtomicMutations.java  1
- graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcherTest.java  1
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6351464&bid=20624614)

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java  5
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/CassandraFetcher.java  2
- graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcherTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6351464&bid=20624614)

**Comment by jorgebay** [comment]
Rebased against `master` and using the mutation options for the batch execution.

I'll add a couple of more tests related to the options.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java  4
- graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcherTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6351464&bid=20624614)

**Comment by jorgebay** [comment]
... and that's it, it should be ready for review.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java  4
- graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcherTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6351464&bid=20624614)

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java [hunk 1]
```diff
@@ -97,4 +112,58 @@ private boolean hasJWTToken() {
...
+  }
+
+  /**
+   * Encapsulates logic to add multiple statements contained in the same operation that need to be
+   * executed in a batch.
+   */
+  public static class BatchContext {
+    private final List<String> statements = Collections.synchronizedList(new ArrayList<>());
+    private final AtomicInteger statementCounter = new AtomicInteger();
+    private final CompletableFuture<ResultSet> future = new CompletableFuture<>();
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:127]: nit: maybe we can name this according to its usage? We know that this is a future from its type

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java [hunk 2]
```diff
@@ -97,4 +112,58 @@ private boolean hasJWTToken() {
...
+          });
+    }
+
+    public void setExecutionResult(Exception ex) {
+      future.completeExceptionally(ex);
+    }
+
+    public int add(String query) {
+      statements.add(query);
+      return statementCounter.incrementAndGet();
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:156]: Why we cannot just return `statements.size()`? I think that we don't need to have the `statementCounter`

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java [hunk 3]
```diff
@@ -20,11 +41,75 @@ protected MutationFetcher(
...
+      if (counter > 1) {
+        // DataStore can be set at most once.
+        // The instance used is the one that came from the user options (if any).
+        buildException =
+            new GraphQLException(
+                "options can only de defined once in an @atomic mutation selection");
+      }
+    }
+
+    if (buildException != null) {
```
Comment (tomekl007) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java:101]: with the current logic, if there is one `buildException` we are failing completing exceptionally with only one exception, right?
Shouldn't we include all possible exceptions, add those as suppressed one and complete exceptionally with all details?

File: graphqlapi/src/test/java/io/stargate/graphql/schema/GraphQlTestBase.java [hunk 4]
```diff
@@ -81,27 +69,66 @@ public void setupEnvironment() {
...
+    } catch (Exception e) {
+      fail("Unexpected exception while mocking dataStore", e);
+    }
 
     graphQlSchema = createGraphQlSchema();
-    graphQl = GraphQL.newGraphQL(graphQlSchema).build();
+    graphQl =
+        GraphQL.newGraphQL(graphQlSchema)
+            // Use parallel execution strategy for mutations (serial is default)
+            .mutationExecutionStrategy(new AsyncExecutionStrategy())
```
Comment (tomekl007) [inline graphqlapi/src/test/java/io/stargate/graphql/schema/GraphQlTestBase.java:105]: should we test both? `AsyncExecutionStrategy` and serial? Maybe a parameterized test?

File: testing/src/main/graphql/betterbotz/schema.json [hunk 5]
```diff
@@ -2088,8 +2088,7 @@
                 "kind": "SCALAR",
                 "name": "Int",
                 "ofType": null
-              },
-              "defaultValue": "-1"
```
Comment (tomekl007) [inline testing/src/main/graphql/betterbotz/schema.json:2092]: why this change?

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java (hunk 2 above)
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:156]: It's a synchronized list, so each call will require a lock acquisition.

I think I'll change it to a synchronized block instead.

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java [hunk 6]
```diff
@@ -97,4 +112,58 @@ private boolean hasJWTToken() {
...
+    }
+
+    public int setDataStore(DataStore dataStore) {
+      this.dataStore = dataStore;
+      return dataStoreCounter.incrementAndGet();
+    }
+
+    public DataStore getDataStoreOrDefault(DataStore defaultDataStore) {
```
[the remark this answers is not in the document]
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:164]: Good call, I'll change it.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java [hunk 7]
```diff
@@ -20,11 +41,75 @@ protected MutationFetcher(
...
+      DataFetchingEnvironment environment,
+      DataStore dataStore,
+      String statement,
+      Exception buildException) {
+    int selections = environment.getOperationDefinition().getSelectionSet().getSelections().size();
+    HTTPAwareContextImpl context = environment.getContext();
+    HTTPAwareContextImpl.BatchContext batchContext = context.getBatchContext();
+
+    if (environment.getArgument("options") != null) {
+      // Users should specify query options once in the batch
```
[the remark this answers is not in the document]
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java:89]: Type validation occurs at GraphQL parsing layer, so on a nullable value, you can expect null or the actual correct type.

These options are used on the `CassandraFetcher` to build the datastore instance.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java (hunk 3 above)
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java:101]: Good question.
In GraphQL, each mutation "selection", should have an individual response. As this is all or nothing, we fail all the selections at the same time with the same error.

Otherwise, it might lead the user to think that one succeeded and the others don't.

File: graphqlapi/src/test/java/io/stargate/graphql/schema/GraphQlTestBase.java (hunk 4 above)
Comment (jorgebay) [reply graphqlapi/src/test/java/io/stargate/graphql/schema/GraphQlTestBase.java:105]: No, we are forcing `AsyncExecutionStrategy` for all DML mutations (see `CustomGraphQLServlet`), so only that strategy will be used for DML mutations.

File: testing/src/main/graphql/betterbotz/schema.json (hunk 5 above)
Comment (jorgebay) [reply testing/src/main/graphql/betterbotz/schema.json:2092]: Good question!

There was a bug in `MutationOptions` where the default value for ttl property was `-1`. But then, `buildStatement()` implementations used `options.get("ttl") != null` condition to check whether ttl was present. So, when the options where defined but the TTL was not set, it would fail by setting TTL to -1.

**Comment by jorgebay** [comment]
Thanks @tomekl007 for the good feedback, I think I'll addressed the issues.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcherTest.java  1
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6351464&bid=20624614)

**Review by dougwettlaufer — APPROVED** [review]

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java [hunk 8]
```diff
@@ -20,11 +41,75 @@ protected MutationFetcher(
...
+      int counter = batchContext.setDataStore(dataStore);
```
Comment (mpenick) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java:90]: The counter value itself isn't ever used. Maybe it would be clearer if this returned boolean value or threw a checked exception: "Datastore already set".

Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java:90]: Good point, I'll fix it.

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java [hunk 9]
```diff
@@ -34,6 +40,14 @@
   private HttpServletRequest request;
   private HttpServletResponse response;
 
+  // We need to manually maintain state between multiple selections in a single mutation
+  // operation to execute them as a batch.
+  // Currently graphql-java batching support is only restricted to queries and not mutations
+  // See https://www.graphql-java.com/documentation/v15/batching/ and
+  // https://github.com/graphql-java/graphql-java/blob/v15.0/src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java#L112-L115
+  // For more information.
+  private final BatchContext batchContext = new BatchContext();
```
Comment (mpenick) [inline graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:49]: This is a pretty complex object (with it allocating many objects itself) to be allocating for non-batch related operations. I wonder if there's a good way to only incur the expense for batch operations.

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java [hunk 10]
```diff
@@ -97,4 +111,61 @@ private boolean hasJWTToken() {
...
+    return batchContext;
+  }
+
+  /**
+   * Encapsulates logic to add multiple statements contained in the same operation that need to be
+   * executed in a batch.
+   */
+  public static class BatchContext {
+    private final List<String> statements = new ArrayList<>();
+    private final Object statementsLock = new Object();
```
Comment (mpenick) [inline graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:125]: This can probably use the monitor of `BatchStatement` w/o requiring another lock object.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java (hunk 8 above)
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java]: Just pushed a fix.

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcherTest.java  1
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6351464&bid=20624614)

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java [hunk 11]
```diff
@@ -97,4 +111,61 @@ private boolean hasJWTToken() {
...
+    public int add(String query) {
+      synchronized (statementsLock) {
+        statements.add(query);
+        return statements.size();
+      }
+    }
+
+    public int setDataStore(DataStore dataStore) {
+      this.dataStore = dataStore;
+      return dataStoreCounter.incrementAndGet();
```
Comment (mpenick) [inline graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java]: This is not likely to be a bottleneck and we're already using synchronization. What about just making this a synchronized method? Then we can remove `volatile` and the atomic counter.

```java 
   // This could return a boolean instead of throwing, but this is an exceptional case and a return value can be accidentally ignored.
    public synchronized void setDataStore(DataStore dataStore) {
      if (this.dataStore != null)
        throw new IllegalStateException("Datastore already set");
      this.dataStore = dataStore;
    }

    public synchronized Optional<DataStore> getDataStore() {
      return Optional.ofNullable(this.dataStore);
    }
```

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java (hunk 9 above)
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:49]: We could use get-or-create semantics but I don't know if its worth the complexity to prevent additional allocations in GraphQL. GraphQL library is already creating a huge context per request...

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java (hunk 10 above)
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:125]: I'm not sure what you mean, we are using strings for now on the graphql side at least until #273 .
In the process, I wanted to leave `BatchContext` implementation as free from data access logic as possible.

Comment (mpenick) [reply graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:125]: IOW, why can't this used `synchronized` methods instead of a lock object?

Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java:125]: got it, I'll change the occurrences to a synchronized method.

Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java]: Done!

File: graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java (hunk 11 above)
Comment (jorgebay) [reply graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java]: I've used a single atomic reference for both.

**Review by mpenick — APPROVED** [review]

**Comment by dougwettlaufer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- graphqlapi/src/test/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcherTest.java  1
- graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6351464&bid=20624614)
